paddle/cinn/ir/group_schedule/tactic/arrange_storage_tactic.cc

```diff
@@ -0,0 +1,407 @@
+  std::unordered_map<ir::Var, ir::Var> cuda_thread_fixed_var_of_load =
+      GetFixedVar(var2for_map, load_block_name, cuda_thread_space_of_load);
+  std::unordered_map<ir::Var, ir::Var> cuda_block_thread_fixed_var_of_store =
+      cuda_block_fixed_var_of_store;
+  cuda_block_thread_fixed_var_of_store.insert(
+      cuda_thread_fixed_var_of_store.begin(),
+      cuda_thread_fixed_var_of_store.end());
+  std::unordered_map<ir::Var, ir::Var> cuda_block_thread_fixed_var_of_load =
+      cuda_block_fixed_var_of_load;
+  cuda_block_thread_fixed_var_of_store.insert(
+      cuda_thread_fixed_var_of_load.begin(),
+      cuda_thread_fixed_var_of_load.end());
+  std::unordered_map<ir::Var, IntSet> store_var_domain =
+      GetVarDomainOfSBlock(var2for_map, store_block_name);
+  std::unordered_map<ir::Var, IntSet> load_var_domain =
+      GetVarDomainOfSBlock(var2for_map, load_block_name);
+  std::vector<ir::Expr> iter_values_of_store =
+      analyzer::GetIterValuesOfAccess(store, store_block);
+  std::vector<ir::Expr> iter_values_of_load =
+      analyzer::GetIterValuesOfAccess(load, load_block);
+  CHECK_EQ(iter_values_of_load.size(), iter_values_of_store.size());
+
+  for (int i = 0; i < iter_values_of_load.size(); ++i) {
+    IntSet block_store_range = Evaluate(iter_values_of_store[i],
+                                        cuda_block_fixed_var_of_store,
+                                        store_var_domain);
+    IntSet block_load_range = Evaluate(
+        iter_values_of_load[i], cuda_block_fixed_var_of_load, load_var_domain);
+    VLOG(6) << "block_store_range of [" << iter_values_of_store[i] << "] = ["
+            << block_store_range.Min() << " : " << block_store_range.Max()
+            << "]";
+    VLOG(6) << "block_load_range of [" << iter_values_of_load[i] << "] = ["
+            << block_load_range.Min() << " : " << block_load_range.Max() << "]";
+    std::optional<bool> is_block_sub_set =
+        block_load_range.ProveSubSet(block_store_range);
+    if (!is_block_sub_set.has_value() || !is_block_sub_set.value()) {
+      VLOG(6) << "load range of a cuda block is not sub set of store";
+      return CudaAxisType::kCudaBlock;
+    }
+
+    IntSet thread_store_range = Evaluate(iter_values_of_store[i],
+                                         cuda_block_thread_fixed_var_of_store,
+                                         store_var_domain);
+    IntSet thread_load_range = Evaluate(iter_values_of_load[i],
+                                        cuda_block_thread_fixed_var_of_load,
+                                        load_var_domain);
+    VLOG(6) << "thread_store_range of [" << iter_values_of_store[i] << "] = ["
+            << thread_store_range.Min() << " : " << thread_store_range.Max()
+            << "]";
+    VLOG(6) << "thread_load_range of [" << iter_values_of_load[i] << "] = ["
+            << thread_load_range.Min() << " : " << thread_load_range.Max()
+            << "]";
+    std::optional<bool> is_thread_sub_set =
```

Comment:
Is this logic strictly right?

Suppose the following kernel code snippet:
```
int N; // tensor length is N
tensor[threadIdx.x] = input[threadIdx.x] + 9
__syncthreads()
output[threadIdx.x] = tensor[(threadIdx.x + 1) % N] // something like shift move
```

It seems the tensor store and load have same IntSet, but we cannot use local 1 in this case. It should be shared, you  have to return CudaAxisType::kCudaThread in this case?

Although, my case may not happen in current CINN, so current logic is fine to me :-) , just a question here, this may not always mathmetically correct?